keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,207 @@
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatch is used to match against the Image URL to determine which images to
+    // provide credentials for.
+    //
+    // The ImageURLMatch string is a URL pattern which can optionally contain a port and a path.  Globs can be used in the hostname (but not the port or the path).
+    //
+    // An image URL matches when it has the same number of parts as the ImageURLMatch, and each part matches.  Additionally the path of ImageURLMatch must be a prefix of the target URL. If the ImageURLMatch contains a port, then the port must match as well.
+    ImageURLMatch []string
+
+    // Exec specifies a custom exec-based plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in YAML and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
+
+We will create new types `RegistryCredentialPluginRequest` and `RegistryCredentialPluginResponse` which will define the interface between the plugin and the kubelet runtime.  After the kubelet matches the image URL to a RegistryCredentialProvider, the kubelet will exec the plugin binary, and pass the JSON encoded request to the plugin via stdin.  This includes the image URL that is to be pulled.  The plugin will report back the response, which includes the credentials that kubelet needs to pull the image.
+
+In the in-tree implementation, the docker keyring, which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.  This struct will be populated by the plugin rather than the in-tree provider.
+
+```go
+// RegistryCredentialPluginRequest is passed to the plugin via stdin, and includes the image that will be pulled by kubelet.
+type RegistryCredentialPluginRequest struct {
+    // Image is used when passed to registry credential providers as part of an image pull
+    // +optional
+    Image string `json:"image,omitempty"`
+}
+
+// RegistryCredentialPluginResponse holds credentials for the kubelet runtime to use for image pulls.  It is returned from the plugin as stdout.
+type RegistryCredentialPluginResponse struct {
+    // +optional
+    ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+
+    // +optional
+    Username *string `json:"username,omitempty"`
+    // +optional
+    Password *string `json:"password,omitempty"`
+
+    // IdentityToken is used to authenticate the user and get
+    // an access token for the registry.
+    // +optional
+    IdentityToken *string `json:"identitytoken,omitempty"`
+
+    // RegistryToken is a bearer token to be sent to a registry
+    // +optional
+    RegistryToken *string `json:"registrytoken,omitempty"`
```

Comment:
If all fields are present, should the kubelet make a request with all the credentials (username, password, identity and registry token)?

Reply:
Drive by comment, since I [dug into this recently](https://github.com/google/go-containerregistry/blob/53ce695e4179e310fd4f7c34099a731829a5d41f/pkg/authn/README.md#the-registry). My understanding of these fields is thus:

`RegistryToken` -> Use this directly as `Bearer` auth against the registry (skip the handshake).
`IdentityToken` -> Use this to perform an [oauth2](https://docs.docker.com/registry/spec/auth/oauth/) handshake.
`Username/Password` -> Use this to perform a [token](https://docs.docker.com/registry/spec/auth/token/) handshake.

This logic doesn't really seem like something the kubelet should handle, though, so it may be sufficient to document that this gets passed through to CRI as [AuthConfig](https://github.com/kubernetes/kubernetes/blob/6ecdd22978783ea51efc34f56f13eb4d9baa446a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L1081-L1092)?

Reply:
I'm in favor of making the kubelet as simple as possible... pass through to CRI auth config sounds good to me.

Would this be usable in combination with dockershim? Given there is work in progress to remove that, I would lean toward saying external credential providers just get used with external container runtimes.

How would this interact with imagePullSecrets? Would the presence of an imagepullsecret override a matching registrycredentialprovider for a given image pull?

Reply:
> Would this be usable in combination with dockershim? Given there is work in progress to remove that, I would lean toward saying external credential providers just get used with external container runtimes.

No plan for dockershim, so that sounds good to me.  

> How would this interact with imagePullSecrets? Would the presence of an imagepullsecret override a matching registrycredentialprovider for a given image pull?

I *think* this is how it works today, with the in-tree providers.  Here, too I was thinking we would model after the current behavior.